File: app/src/main/java/com/example/stramitapp/zebraconnection/connect/ReaderDeviceAdapter.java
package com.example.stramitapp.zebraconnection.connect;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.example.stramitapp.R;
import com.zebra.rfid.api3.ReaderDevice;
import java.util.List;

public class ReaderDeviceAdapter extends RecyclerView.Adapter<ReaderDeviceAdapter.ReaderViewHolder> {
    public interface OnItemClickListener {
        void onItemClick(ReaderDevice device);
    }

    private List<ReaderDevice> readerList;
    private final OnItemClickListener listener;
    private String connectedReaderName = null;

    public ReaderDeviceAdapter(List<ReaderDevice> readerList, OnItemClickListener listener) {
        this.readerList = readerList;
        this.listener = listener;
    }

    public void setReaderList(List<ReaderDevice> readerList) {
        this.readerList = readerList;
        notifyDataSetChanged();
    }

    public void setConnectedReaderName(String name) {
        connectedReaderName = name;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ReaderViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_reader_device, parent, false);
        return new ReaderViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ReaderViewHolder holder, int position) {
        ReaderDevice device = readerList.get(position);
        holder.bind(device);
    }

    @Override
    public int getItemCount() {
        return readerList != null ? readerList.size() : 0;
    }

    class ReaderViewHolder extends RecyclerView.ViewHolder {
        TextView textViewName;
        TextView textViewSerial;
        TextView textViewModel;

        ReaderViewHolder(View itemView) {
            super(itemView);
            textViewName = itemView.findViewById(R.id.textViewReaderName);
            textViewSerial = itemView.findViewById(R.id.textViewReaderSerial);
            textViewModel = itemView.findViewById(R.id.textViewReaderModel);

            itemView.setOnClickListener(v -> {
                int position = getAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    listener.onItemClick(readerList.get(position));
                }
            });
        }

        void bind(ReaderDevice device) {
            textViewName.setText(device.getName());
            textViewSerial.setText("Serial: " + device.getSerialNumber());
            textViewModel.setText("Model: " + device.getName());
        }
    }
}

File: app/src/main/java/com/example/stramitapp/ui/reader_list/ReaderDeviceAdapter.kt
package com.example.stramitapp.ui.reader_list

import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.stramitapp.R
import com.zebra.rfid.api3.ReaderDevice

class ReaderDeviceAdapter(private val listener: OnItemClickListener) :
    ListAdapter<ReaderDevice, ReaderDeviceAdapter.ReaderDeviceViewHolder>(ReaderDeviceDiffCallback()) {

    private var selectedReaderName: String? = null

    fun setSelectedReader(readerName: String?) {
        selectedReaderName = readerName
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int):
            ReaderDeviceViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_reader_device, parent, false)
        return ReaderDeviceViewHolder(view)
    }

    override fun onBindViewHolder(holder: ReaderDeviceViewHolder, position: Int) {
        val readerDevice = getItem(position)
        holder.bind(readerDevice, readerDevice.name == selectedReaderName)
    }

    inner class ReaderDeviceViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {

        private val readerNameTextView: TextView = itemView.findViewById(R.id.textViewReaderName)
        private val readerSerialTextView: TextView = itemView.findViewById(R.id.textViewReaderSerial)
        private val readerModelTextView: TextView = itemView.findViewById(R.id.textViewReaderModel)
        private val rootLayout: View = itemView.findViewById(R.id.reader_box_layout)

        init {
            itemView.setOnClickListener {
                val position = bindingAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    val readerDevice = getItem(position)
                    listener.onItemClick(readerDevice)
                }
            }
        }

        fun bind(readerDevice: ReaderDevice, isSelected: Boolean) {
            readerNameTextView.text = readerDevice.name
            readerSerialTextView.text = readerDevice.serialNumber
            readerModelTextView.text = readerDevice.name

            if (isSelected) {
                rootLayout.setBackgroundResource(R.color.app_blue) // Darker blue when selected
                readerNameTextView.setTextColor(Color.WHITE)
                // Update other text colors if needed
            } else {
                rootLayout.setBackgroundResource(R.color.reader_list_item_blue)
                readerNameTextView.setTextColor(itemView.context.getColor(R.color.text_color_primary))
            }
        }
    }

    interface OnItemClickListener {
        fun onItemClick(readerDevice: ReaderDevice)
    }

    class ReaderDeviceDiffCallback : DiffUtil.ItemCallback<ReaderDevice>() {
        override fun areItemsTheSame(oldItem: ReaderDevice, newItem: ReaderDevice):
                Boolean {
            return oldItem.name == newItem.name
        }

        override fun areContentsTheSame(oldItem: ReaderDevice, newItem: ReaderDevice):
                Boolean {
            return oldItem.name == newItem.name && oldItem.serialNumber == newItem.serialNumber
        }
    }
}
